### PR TITLE
PR: Add initial restart implementation

### LINKED ISSUE #2181: Add a restart option on the file menu
## 🚀 Feature

Add a restart method to the main application and add it to the corresponding menu.

## Motivation

Some configuration options will need a restart so we need to provide a a way to restart

## Pitch

N.A.

## Alternatives

N.A.

## Additional context

This is part of the configuration/settings work.

### PR DESCRIPTION
**Pull request by goanpeca**
# Description

Fixes #2181

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

File: napari/__main__.py [hunk 1]
```diff
@@ -14,18 +14,18 @@
 from textwrap import wrap
 from typing import Any, Dict, List
 
-from . import Viewer, __version__, layers, run, view_path
-from .components.viewer_model import valid_add_kwargs
-from .utils import citation_text, sys_info
-from .utils.settings import SETTINGS
+from napari import Viewer, __version__, layers, run, view_path
```
Comment (sofroniewn) [inline napari/__main__.py:17]: I don't mind, but was there a reason to switch to absolute imports in main? I believe @jni had once said something about preferring relative, but i might be wrong ...

**Review by sofroniewn — COMMENTED** [review]
I was trying to use this now, but I'm noticing something a little funky going on.

If I launch napari with `python examples/add_image.py` then I get

Screen Shot 2021-04-12 at 9 02 26 AM

If I launch napari just using `napari` then I get

Screen Shot 2021-04-12 at 9 03 15 AM

Notice how I only have the "restart" menu option in the 2nd case, and also that the dock widgets have labels only in the second case, and I only have the sample data menubar in the 2nd case (cc @tlambert03).

If I click restart I then get the following segfault

```
/Users/nsofroniew/opt/anaconda3/bin/pythonw: line 3: 10253 Segmentation fault: 11  /Users/nsofroniew/opt/anaconda3/python.app/Contents/MacOS/python "$@"
```

If I launch with `pythonw examples/add_image.py` then things look like the first case, so it's not a pythonw vs python difference but a launching via `napari` vs running a script difference. Ideally I think we want to make these as similar as possible if not identical ways to run napari.

**Comment by sofroniewn** [comment]
Ok, my comment above with screenshots was some weirdness happening in my environment - I have a suspicion that i installed a napari plugin that pinned napari to a version and that messed everything up. I'll investigate that more elsewhere.

Now though when i click the restart button I still don't see napari relaunch though. But that could still be on my end ...

File: napari/__main__.py (hunk 1 above)
Comment (goanpeca) [reply napari/__main__.py:17]: It was not working for me with relative imports so I changed them to see what could be the problem.

**Comment by goanpeca** [comment]
>Now though when i click the restart button I still don't see napari relaunch though. But that could still be on my end ...

I may need to do some extra things for mac bundled apps.

This is more of a POC to see what needs to be fixed, so if you could test on a fresh env that would be great @sofroniewn !

File: napari/__main__.py (hunk 1 above)
Comment (goanpeca) [reply napari/__main__.py:17]: @sofroniewn on master what do you get when running 

`python napari/__main__.py` I may be able to have a workaround. Will explore that so I can revert to relative imports.

**Comment by sofroniewn** [comment]
> This is more of a POC to see what needs to be fixed, so if you could test on a fresh env that would be great @sofroniewn !

Clean environment on my mac, restart doesn't work for me with either PySide2/ PyQt5. napari closes, but it never re-opens :-)

**Comment by goanpeca** [comment]
Thanks :), will take a look again :)

**Comment by tlambert03** [comment]
> restart doesn't work for me with either PySide2/ PyQt5. napari closes, but it never re-opens :-)

same for me.  having a look now (I vaguely recall this being challenging when I was trying to implement for plugins before)

File: napari/__main__.py (hunk 1 above)
Comment (sofroniewn) [reply napari/__main__.py:17]: no luck restarting with `python napari/__main__.py` either

**Comment by tlambert03** [comment]
I think you need to actually start the `QProcess` right?
works for me if I start it (have to startDetached) as such:
```python
        process.setProgram(sys.executable)
        process.setArguments(sys.argv)
        process.startDetached()
        self.close(quit_app=True)
```

**Comment by goanpeca** [comment]
😆 , I forgot to push that 😆

File: napari/_qt/qt_main_window.py [hunk 2]
```diff
@@ -222,6 +223,14 @@ def closeEvent(self, event):
 
         event.accept()
 
+    def restart(self):
+        """Restart the napari application in a detached process."""
+        process = QProcess()
+        # For the Mac App this might need to be different
+        process.setProgram(sys.executable)
+        process.setArguments(sys.argv)
```
Comment (tlambert03) [inline napari/_qt/qt_main_window.py]:
```suggestion
        process.setArguments(sys.argv)
        process.startDetached()
```

File: napari/_qt/qt_main_window.py [hunk 3]
```diff
@@ -430,6 +439,10 @@ def _add_file_menu(self):
             lambda: self._qt_window.close(quit_app=True)
         )
 
+        restartAction = QAction(trans._('Restart'), self._qt_window)
+        restartAction.setShortcut('Ctrl+R')
```
Comment (tlambert03) [inline napari/_qt/qt_main_window.py]: I'm not sure we need a shortcut for this one... we already have one "immediately loose your work" short cut (`Ctrl+Q`) 😄 ... and we don't yet double check that they are ok with closing.  Ok to leave without a shortcut?

**Review by tlambert03 — COMMENTED** [review]
I'm generally ok with this.

I don't think it's going to work with relative imports in `__main__.py` though, since most of the time it will be trying to execute `python napari/__main__.py` ... which doesn't work (without the `-m` flag).  so we either need to do some `sys.argv` parsing in the restart logic, and perhaps hardcode the `-m napari` ... or use absolute imports in `__main__.py` (which I am personally fine with)

**Comment by tlambert03** [comment]
also, I think we're going to need a little more context-awareness.  If you start a viewer in IPython and use this, it will just kill your ipython session but not restart (unsurprisingly):
```python
In [2]: import napari

In [3]: v = napari.Viewer()

In [4]: Python 3.9.2 | packaged by conda-forge | (default, Feb 21 2021, 05:02:20)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.22.0 -- An enhanced Interactive Python. Type '?' for help.

WARNING: your terminal doesn't support cursor position requests (CPR).
```

File: napari/_qt/qt_main_window.py (hunk 3 above)
Comment (goanpeca) [reply napari/_qt/qt_main_window.py]: This was for quick testing (and is also overriding the refresh one on the viewer) so I intend to remove it :)

**Comment by goanpeca** [comment]
> If you start a viewer in IPython and use this, it will just kill your ipython session but not restart (unsurprisingly):

Thanks for flagging that indeed, the feedback I needed :-p

File: napari/_qt/qt_main_window.py (hunk 3 above)
Comment (tlambert03) [reply napari/_qt/qt_main_window.py]: >  overriding the refresh one on the viewer

oh right! :)

**Comment by tlambert03** [comment]
the big benefit of this I think will be for the bundled app... so we might even consider adding it only when `utils.misc.running_as_bundled_app` is True?  I know that leaves out the very reasonable case of installing a language pack or something after having started napari from the command line, for instance.  But since the new process doesn't start in the same terminal window (it's just some ghost background process), it's a little odd even in that scenario.  what do you think?

**Comment by goanpeca** [comment]
> But since the new process doesn't start in the same terminal window (it's just some ghost background process), it's a little odd even in that scenario. what do you think?

Yeah, it is!

>the big benefit of this I think will be for the bundled app... so we might even consider adding it only when utils.misc.running_as_bundled_app is True?

Agreed!

I think we might be able to do the translation in place with the new TranslationString thinguie that defers translations. Will require some fiddling with widgets, but I think we can make it work eventually

**Comment by goanpeca** [comment]
Any pointers on creating the bundled app locally so I can test this @tlambert03 @sofroniewn ?

**Comment by tlambert03** [comment]
> Any pointers on creating the bundled app locally

should be able to do this:
```
pip install briefcase tomlkit wheel dmgbuild
python bundle.py
```

**Comment by tlambert03** [comment]
also, a tip that may not be immediately obvious, particularly if you have experience with pyinstaller:  briefcase bundles a literal "unfrozen" python environment with all of the source code.  So once you've built something, rather than changing the source code and rebuilding again, you can just go into the build application, edit the source (for instance, on a mac, at `napari.app/Contents/Resources/app/napari`), and then restart the application (no rebuild necessary)

**Comment by goanpeca** [comment]
>also, a tip that may not be immediately obvious, particularly if you have experience with pyinstaller: briefcase bundles a literal "unfrozen" python environment with all of the source code. So once you've built something, rather than changing the source code and rebuilding again, you can just go into the build application, edit the source (for instance, on a mac, at napari.app/Contents/Resources/app/napari), and then restart the application (no rebuild necessary)


Great! thanks :)

File: napari/__main__.py (hunk 1 above)
Comment (goanpeca) [reply napari/__main__.py:17]: Please try with the bundled app @sofroniewn  :)

**Comment by goanpeca** [comment]
@tlambert03 the menu is now only added on the mac bundle. I also left some methods that we could use if we need to trigger a restart eventually.

¿Could you try it? 

Thanks!

**Comment by tlambert03** [comment]
it works!  lovely

**Review by sofroniewn — APPROVED** [review]
Ok nice, this now looks good to go from me!!

**Pull request merged**